### ISSUE TITLE
Using ChainedScheduler with ReduceLROnPlateau leads to unexpected keyword argument error

### ISSUE THREAD

**Issue by amueller** [pytorch/pytorch#110761]
### 🚀 The feature, motivation and pitch

I want to use ReduceLROnPlateau together with another schedule, say a warm-up or cosine schedule.
My understanding is that the suggested way to do that is through ``ChainedScheduler``.
However, when I try to call ``step`` with a validation loss, I get a ``TypeError``, since ``ChainedScheduler`` doesn't accept ``kwargs`` in ``step``.
I'm still pretty new to pytorch, so please let me know if I'm overlooking some other solution.

I don't see a super obvious solution. One would be to pass all ``kwargs`` to all steps in the chain, but that would mean all the schedulers would have to accept arbitrary ``kwargs`` which would hide errors. The other option that comes to mind would be to use inspect to figure out which parameters to pass to which scheduler. Or the ``step`` of ``ChainedScheduler`` could take a dict mapping from schedulers to parameters, but that seems more complicated than just calling all the schedules explicitly and not using ``ChainedScheduler``.

**Comment by jon-chuang** [comment]
Could you provide a small reproducer script?

**Comment by janeyx99** [comment]
I am not sure if this is your desired chaining, but you should be able to register two schedulers on one optimizer, something like:
```
schedulers = []
optimizer = ...
schedulers[0] = ReduceLROnPlateau(
    optimizer,
    patience=5,
    cooldown=0,
    threshold_mode="abs",
    mode="min",
    threshold=0.1,
)
schedulers[1] = MultiStepLR(optimizer gamma=0.1, milestones=[3, 8, 12])
for s in schedulers:
    s.step(...)
```

**Comment by amueller** [comment]
Thank you @janeyx99. A work-around like that would work, though you can't write the loop like this, since ``MultiStepLR`` doesn't take any arguments in ``step`` so a similar problem remains. You can explicitly call them individually, though.

Doesn't work:
```python
from torch.optim.lr_scheduler import CosineAnnealingLR, ReduceLROnPlateau, ChainedScheduler
from torch.nn import Parameter
from torch.optim import SGD
import torch

import numpy as np
model = [Parameter(torch.randn(2, 2, requires_grad=True))]
optimizer = SGD(model, 0.1)

max_epochs = 100
cosine_scheduler = CosineAnnealingLR(optimizer, T_max=max_epochs)
reduce_scheduler = ReduceLROnPlateau(optimizer)
chained_scheduler = ChainedScheduler([cosine_scheduler, reduce_scheduler])
lrs = []
losses = []

for epoch in range(100):
    loss = 1 + np.exp(-epoch / 10)    
    lrs.append([group['lr'] for group in optimizer.param_groups][0])
    optimizer.step()
    chained_scheduler.step(metrics=loss)
    losses.append(loss)
```

Doesn't work:
```python
from torch.optim.lr_scheduler import CosineAnnealingLR, ReduceLROnPlateau
from torch.nn import Parameter
from torch.optim import SGD
import torch
model = [Parameter(torch.randn(2, 2, requires_grad=True))]
optimizer = SGD(model, 0.1)

max_epochs = 100
cosine_scheduler = CosineAnnealingLR(optimizer, T_max=max_epochs)
reduce_scheduler = ReduceLROnPlateau(optimizer)
lrs = []
losses = []
schedulers = [cosine_scheduler, reduce_scheduler]

for epoch in range(100):
    loss = 1 + np.exp(-epoch / 10)    
    lrs.append([group['lr'] for group in optimizer.param_groups][0])
    optimizer.step()
    for scheduler in schedulers:
        scheduler.step(loss)
    losses.append(loss)
```  

Works:

```python
from torch.optim.lr_scheduler import CosineAnnealingLR, ReduceLROnPlateau
from torch.nn import Parameter
from torch.optim import SGD
import torch

import numpy as np
model = [Parameter(torch.randn(2, 2, requires_grad=True))]
optimizer = SGD(model, 0.1)

max_epochs = 100
cosine_scheduler = CosineAnnealingLR(optimizer, T_max=max_epochs)
reduce_scheduler = ReduceLROnPlateau(optimizer)
lrs = []
losses = []

for epoch in range(100):
    loss = 1 + np.exp(-epoch / 10)    
    lrs.append([group['lr'] for group in optimizer.param_groups][0])
    optimizer.step()
    cosine_scheduler.step()
    reduce_scheduler.step(metrics=loss)
    losses.append(loss)
```

**Comment by janeyx99** [comment]
Oh right, yes, I forgot that their step takes different arguments. Does it suffice for your use case to not call step in a loop?

**Comment by amueller** [comment]
Sure my solution is fine, I just thought it odd that ``ChainedScheduler`` doesn't allow you to chain arbitrary schedulers. That seems a bit surprising design-wise. But I can certainly make it work without, so feel free to close if you don't think that's an issue.

**Comment by janeyx99** [comment]
Ah I see. I do agree that our LR Schedulers are not in their best shape (there hasn't been any active dedicated maintenance for it for a while). I will leave this open as a discussion for how to best design ChainedScheduler. I'm actually curious why we even have a ChainedScheduler API when one can register multiple LR Schedulers already.

**Comment by amueller** [comment]
I think it helps with making serialization more modular? With chaining I can re-use the same serialization and deserialization code no matter how many schedulers I use. I guess I could always store a list, but having the training state as model state + optimizer state + scheduler state seem convenient? Though I am still really new to pytorch so I know nothing.

**Comment by sergiev** [comment]
@amueller thank you so much for providing example that works! You saved a lot of time to folks like me ❤️

**Comment by pupeno** [comment]
I just provided a couple of PRs that would also cover this issue. The relevant comment is here: https://github.com/pytorch/pytorch/issues/68978#issuecomment-5500830296
